kbs/src/attestation/backend.rs

```diff
@@ -0,0 +1,362 @@
+                Arc::new(grpc_coco_as) as _
+            }
+            #[cfg(feature = "intel-trust-authority-as")]
+            AttestationServiceConfig::IntelTA(cfg) => {
+                let intel_ta = super::intel_trust_authority::IntelTrustAuthority::new(cfg)
+                    .await
+                    .map_err(|e| Error::AttestationServiceInitialization { source: e })?;
+                Arc::new(intel_ta) as _
+            }
+        };
+
+        let session_map = Arc::new(SessionMap::new());
+
+        tokio::spawn({
+            let session_map_clone = session_map.clone();
+            async move {
+                loop {
+                    tokio::time::sleep(std::time::Duration::from_secs(60)).await;
+                    session_map_clone
+                        .sessions
+                        .retain_async(|_, v| !v.is_expired())
+                        .await;
+                }
+            }
+        });
+        Ok(Self {
+            inner,
+            timeout: config.timeout,
+            session_map,
+        })
+    }
+
+    pub async fn set_policy(&self, request: &[u8]) -> Result<()> {
+        self.__set_policy(request)
+            .await
+            .map_err(|e| Error::SetPolicy { source: e })
+    }
+
+    async fn __set_policy(&self, request: &[u8]) -> anyhow::Result<()> {
+        let input: SetPolicyInput =
+            serde_json::from_slice(request).context("parse set policy request")?;
+        self.inner.set_policy(&input.policy_id, &input.policy).await
+    }
+
+    pub async fn auth(&self, request: &[u8]) -> Result<HttpResponse> {
+        self.__auth(request)
+            .await
+            .map_err(|e| Error::RcarAuthFailed { source: e })
+    }
+
+    async fn __auth(&self, request: &[u8]) -> anyhow::Result<HttpResponse> {
+        let request: Request = serde_json::from_slice(request).context("deserialize Request")?;
+        let version = Version::parse(&request.version).context("failed to parse KBS version")?;
+        if !VERSION_REQ.matches(&version) {
+            bail!(
+                "expected version: {}, requested version: {}",
+                *VERSION_REQ,
+                request.version
+            );
+        }
+
+        let challenge = self
+            .inner
+            .generate_challenge(request.tee, request.extra_params.clone())
+            .await
+            .context("generate challenge")?;
+
+        let session = SessionStatus::auth(request, self.timeout, challenge).context("Session")?;
+
+        let response = HttpResponse::Ok()
+            .cookie(session.cookie())
+            .json(session.challenge());
+
+        self.session_map.insert(session);
+
+        Ok(response)
+    }
+
+    pub async fn attest(&self, attestation: &[u8], request: HttpRequest) -> Result<HttpResponse> {
+        self.__attest(attestation, request)
+            .await
+            .map_err(|e| Error::RcarAttestFailed { source: e })
+    }
+
+    async fn __attest(
```

Comment:
this convention of creating "private functions w/ underscore prefixes doesn't look like idiomatic rust. can we inline `__myfunc()` in `myfunc()` or if they need to be separated can we put it into a module?

Reply:
Do you mean to use `#[inline]` upon `pub async fn attest(...)`?

Reply:
no, just move the __attest code into the attest function (unless there is a reason to keep them apart, then we can move it it into a module maybe)

Reply:
Oh. The wrapping is for error conversion. If move all logic from `__attest` to `attest`, we need a lot of `map_err` from `anyhow::Error` to `Error`

Reply:
Yes. The `#[from]` could automatically convert *ALL* `anyhow:Error` to `Error` with `?`. In this case we want to do some classfication upon the different `anyhow::Error` due to `Auth` and `Attest` of RCAR

Reply:
ok, if we want to mix anyhow and concrete error types we have to resort to gymnastics like this function wrapping. If we bother about the concrete errors I'd suggest declaring them explicitly in an error enum, at least in new code. so instead of 

```rust
request.cookie(KBS_SESSION_ID).context("cookie not found")?;
```

```rust
#[error("cookie not found")]
CookieNotFound,
...
request.cookie(KBS_SESSION_ID).ok_or(Error::CookieNotFound)?;
```

Reply:
Right. My original aim is to avoid too many error types. Overall, they are only `RCAR attest failed`, because of an `anyhow::Error` and `RCAR auth failed`, because of an `anyhow::Error`. The concrete error inside could be all coverted by `anyhow`. I like the concise error type definition style myself